File: XamarinStore.iOS/Views/StringUIPicker.cs
using System;
using System.Linq;
using System.Collections.Generic;
using MonoTouch.UIKit;
using System.Drawing;

namespace XamarinStore
{
	public class StringUIPicker : UIPickerView
	{
		public event EventHandler SelectedItemChanged;
		public StringUIPicker ()
		{
		}
		string[] items;
		public IEnumerable<string> Items
		{
			get{ return items; }
			set{ 
				items = value.ToArray ();
				Model = new PickerModel {
					Items = items,
					Parent = this,
				};
			}
		}
		int currentIndex;
		public int SelectedIndex {
			get{ return currentIndex; }
			set{
				if (currentIndex == value)
					return;
				currentIndex = value;
				this.Select (currentIndex, 0, true);
				if (SelectedItemChanged != null)
					SelectedItemChanged (this, EventArgs.Empty);
			}
		}
		public string SelectedItem
		{
			get { 
				return items.Length <= currentIndex ? "" : items [currentIndex];
			}
			set {
				if(!items.Contains(value))
					return;
				currentIndex = Array.IndexOf (items, value);
			}
		}

		UIActionSheet sheet;
		public void ShowPicker(UIView viewForPicker)
		{
			sheet = new UIActionSheet();

			sheet.AddSubview(this);

			var toolbarPicker = new UIToolbar (new RectangleF (0, 0, viewForPicker.Frame.Width, 44)) {
				Items = new UIBarButtonItem[] {
					new UIBarButtonItem (UIBarButtonSystemItem.FlexibleSpace), 
					new UIBarButtonItem (UIBarButtonSystemItem.Done, (sender, args) => sheet.DismissWithClickedButtonIndex (0, true)), 
				},
				BarTintColor = this.BackgroundColor,
			};

			sheet.AddSubviews(toolbarPicker);

			sheet.BackgroundColor = UIColor.Clear;
			sheet.ShowInView(viewForPicker);
			UIView.Animate(.25, () => sheet.Bounds = new RectangleF (0, 0, viewForPicker.Frame.Width, 485));

		}


		class PickerModel : UIPickerViewModel
		{
			public StringUIPicker Parent { get; set; }
			public string[] Items = new string[0];

			public override int GetComponentCount (UIPickerView picker)
			{
				return 1;
			}

			public override int GetRowsInComponent (UIPickerView picker, int component)
			{
				return Items.Length;
			}

			public override string GetTitle (UIPickerView picker, int row, int component)
			{

				return Items [row];
			}

			public override void Selected (UIPickerView picker, int row, int component)
			{
				if (Parent != null)
					Parent.SelectedIndex = row;
			}

		}

	}
}

File: F#/XamarinStore.iOS/Views/StringUIPicker.fs
namespace XamarinStore

open System
open System.Linq
open System.Collections.Generic
open MonoTouch.UIKit
open System.Drawing

type PickerModel() =
    inherit UIPickerViewModel()

    member val Parent: StringUIPicker = null with get, set
    member val Items: string array = [||] with get, set

    override this.GetComponentCount picker =
        1

    override this.GetRowsInComponent (picker, ``component``) =
        this.Items.Length

    override this.GetTitle (picker, row, ``component``) =
        this.Items.[row]

    override this.Selected (picker, row, ``component``) =
        if this.Parent <> null then
            this.Parent.SelectedIndex <- row

and [<AllowNullLiteralAttribute>] StringUIPicker() =
    inherit UIPickerView()

    let mutable currentIndex = 0
    let mutable items :string array = [||]
    let mutable sheet : UIActionSheet = null

    member this.Items
        with get () = items :> string seq
        and set (value:string seq) = items <- value.ToArray()
                                     this.Model <- new PickerModel(Items = items, Parent = this)

    member val SelectedItemChanged = fun x->() with get,set

    member this.SelectedIndex
        with get () = currentIndex
        and set value =
            if currentIndex <> value then
                currentIndex <- value
                this.Select (currentIndex, 0, true)
                this.SelectedItemChanged this

    member this.SelectedItem
        with get () = if items.Length <= currentIndex then "" else items.[currentIndex]
        and set value = if items.Contains(value) then
                            currentIndex <- Array.FindIndex(items, fun x-> x = value)

    member this.ShowPicker(viewForPicker : UIView) =
        sheet <- new UIActionSheet()

        sheet.AddSubview this

        let doneButton = new UIBarButtonItem(UIBarButtonSystemItem.Done)
        doneButton.Clicked.Add(fun _ -> sheet.DismissWithClickedButtonIndex (0, true))
        let toolbarPicker = new UIToolbar (new RectangleF (0.0f, 0.0f, viewForPicker.Frame.Width, 44.0f))
        toolbarPicker.Items <- [|new UIBarButtonItem (UIBarButtonSystemItem.FlexibleSpace); doneButton|]
        toolbarPicker.BarTintColor <- this.BackgroundColor

        sheet.AddSubviews toolbarPicker

        sheet.BackgroundColor <- UIColor.Clear
        sheet.ShowInView viewForPicker
        UIView.Animate(0.25, fun () -> sheet.Bounds <- new RectangleF (0.0f, 0.0f, viewForPicker.Frame.Width, 485.0f))